server-readonly/src/main/java/org/eclipse/rdf4j/http/server/readonly/QueryResponder.java

```diff
@@ -0,0 +1,198 @@
+import org.eclipse.rdf4j.query.resultio.BooleanQueryResultWriterRegistry;
+import org.eclipse.rdf4j.query.resultio.QueryResultFormat;
+import org.eclipse.rdf4j.query.resultio.QueryResultIO;
+import org.eclipse.rdf4j.query.resultio.TupleQueryResultFormat;
+import org.eclipse.rdf4j.repository.Repository;
+import org.eclipse.rdf4j.repository.RepositoryConnection;
+import org.eclipse.rdf4j.rio.RDFFormat;
+import org.eclipse.rdf4j.rio.RDFHandlerException;
+import org.eclipse.rdf4j.rio.Rio;
+import org.eclipse.rdf4j.rio.UnsupportedRDFormatException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.RequestHeader;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class QueryResponder {
+	/**
+	 * The repository that is being served.
+	 */
+	@Autowired
+	private final Repository repository;
+
+	@Autowired
+	public QueryResponder(Repository repository) {
+		this.repository = repository;
+	}
+
+	@RequestMapping(value = "/sparql", method = RequestMethod.POST, consumes = APPLICATION_FORM_URLENCODED_VALUE)
+	public void sparqlPostURLencoded(
+			@RequestParam(value = "default-graph-uri", required = false) String defaultGraphUri,
+			@RequestParam(value = "named-graph-uri", required = false) String namedGraphUri,
+			@RequestParam(value = "query") String query,
+			@RequestHeader(ACCEPT) String acceptHeader,
+			HttpServletRequest request, HttpServletResponse response) throws IOException {
+		doSparql(request, query, acceptHeader, defaultGraphUri, namedGraphUri, response);
+	}
+
+	@RequestMapping(value = "/sparql", method = RequestMethod.GET)
+	public void sparqlGet(
+			@RequestParam(value = "default-graph-uri", required = false) String defaultGraphUri,
+			@RequestParam(value = "named-graph-uri", required = false) String namedGraphUri,
+			@RequestParam(value = "query") String query,
+			@RequestHeader(ACCEPT) String acceptHeader,
+			HttpServletRequest request, HttpServletResponse response) throws IOException {
+		doSparql(request, query, acceptHeader, defaultGraphUri, namedGraphUri, response);
+	}
+
+	private void doSparql(HttpServletRequest request, String query, String acceptHeader, String defaultGraphUri,
+			String namedGraphUri, HttpServletResponse response) throws IOException {
+
+		try (RepositoryConnection connection = repository.getConnection()) {
+			Query preparedQuery = connection.prepareQuery(QueryLanguage.SPARQL, query);
+			for (QueryTypes qt : QueryTypes.values()) {
+				if (qt.accepts(preparedQuery, acceptHeader)) {
+					qt.evaluate(preparedQuery, acceptHeader, response);
+				}
+			}
+		} catch (MalformedQueryException | MismatchingAcceptHeaderException mqe) {
+			response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+		}
+	}
```

Comment:
Is defaultGraphUri or namedGraphUri supported here?

Reply:
It was not, I think I figured the behaviour out in a804d5ad11a1d95556bbc7471c408c62654c7eba but is untested as of now.